### ISSUE TITLE
[Feature request] In browser printing

### ISSUE THREAD

**Issue by ryno1337** [mozilla-mobile/fenix#10217]
### Why/User Benefit/User Problem
so user can print from mobile browser, there is no print option in browser

### What/Requirements
add print option to browser menu

### Acceptance Criteria (how do I know when I’m done?)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-264)

**Comment by Gabriel85** [comment]
Any progress on that?
It a very important feature for people without a PC (like my 85 year old grandma). Plus all the big boys (like Chrome and Samsung Browser and even Fennec) support it! I'm sure it is a dealbreaker for many people..

**Comment by mohibo** [comment]
any progress?

**Comment by zilexa** [comment]
I am a Firefox user since v0.6, I also use it on Android and run my own Firefox Sync server.
I understand and I am grateful a new Android version had to be build from the ground up.

But a year since I switched to the new Firefox for Android and still impossible to print! 

MOZILLA, are you aware how many people shop online? Probably all your users. And they return lots of stuff. You need to be able to print.

The fact this still isn't possible makes me want to break with Firefox and Mozilla completely. It's incredible.

**Comment by ueen** [comment]
This is actually hard to belive while other, smaller projects, are able to implement printing https://github.com/CookieJarApps/SmartCookieWeb-Preview

**Comment by zilexa** [comment]
🤦🏾

**Comment by SeburoSUMO** [comment]
User feedback- https://support.mozilla.org/en-US/questions/1344026

**Comment by SeburoSUMO** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1344099

**Comment by lveillard** [comment]
Hello guys! This feature is important, but even more important is to clarify that it is not working.
Right now a button linked to a window.print() action will just do nothing in Firefox. Would be nice to at least show a little alert by default "sorry, printing is not compatible with Firefox" 

Thanks!

**Comment by eUgEntOptIc44** [comment]
> Hello guys! This feature is important, but even more important is to clarify that it is not working. Right now a button linked to a window.print() action will just do nothing in Firefox. Would be nice to at least show a little alert by default "sorry, printing is not compatible with Firefox"

first of all being a fronted dev myself I agree with @lveillard we need a feedback for the user concering the lack of th e print function to not get blamed for it ourselves.

-----------

Second: is there any progress on this issue?

To me (as a longtime firefox user) it looks there's very little to no progress on the topics:

* **keyboard shortcuts** ([mozilla-mobile/android-components -> #1978](https://github.com/mozilla-mobile/android-components/issues/1978), #3729 
* **printing** (#10217) and **print to pdf** (#3709)

I'd like to point out that this **highly frustrating** for me personally. I need to use a chromium based browser on my tablet obviously lacking firefox sync and uBlock Origin as features.

@eliserichards has done the engineering triage in April 2020 more than 1.5 years ago. But since then I haven't heard of any progress on this.

**Comment by SeburoSUMO** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1357316

**Comment by aquatic7** [comment]
This is why you move from Firefox to another browser on Android. They simply just don't care about their 'customers'.

**Comment by SeburoSUMO** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1358276

**Comment by SeburoSUMO** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1358490

**Comment by headbank** [comment]
Just wondering, should there be a corresponding ticket on this issue on bugzilla.mozilla.org ? I had a look and can't find anything there relating to this on Android/Fenix. (Mind I'm not clear what exactly is the dividing line of responsibility between these two trackers, but this being quite a core feature I could see it requiring low-level work whose progress might presumably belong over there.)

I wish we had some developer input here giving us some insight as to even just the scope of what's required for this ticket to move on (let alone whether anyone is seriously looking at doing so). For example is any of the Fennec codebase for this at all portable, or must they start from scratch?

**Comment by headbank** [comment]
This [Android Developers page](https://developer.android.com/training/printing/custom-docs) seems relevant. If I follow it correctly, to give a print job to Android's print service, you must give it a PDF file.

If that's correct, then I guess this ticket depends on #3709 ?

**Comment by plwt** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1360517

**Comment by plwt** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1363912

**Comment by headbank** [comment]
I wonder if, using Javascript, one could generate a static copy of the current page and feed it into the [Share API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Share_API)?

It would require external resources like stylesheets and images to be converted to inline formats, but both of these I know are possible. More complex content like canvas/svg might be less so, though.

Since bookmarklets seem to be working again lately, perhaps one could be crafted so the page can be passed to another browser (or other HTML-viewing app) for printing.

Sorry for the digression but just a thought on a possible workaround approach while we wait. I'll try to investigate the possibility myself as time allows.

**Comment by plwt** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1367297

**Comment by plwt** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1369904

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
Would implementation of this provide the webpage to the OS to print with external software, or is this request to add the ability to print to Fenix?

**Comment by aaronwi** [comment]
2 years later and still no printing?

**Comment by benyaminl** [comment]
@aaronwi meanwhile, use SingleHTML save and use nightly... it really powerful for now at least..

**Comment by Erudition** [comment]
Wow

**Comment by plwt** [comment]
User feedback: https://support.mozilla.org/en-US/questions/1380196

**Comment by benyaminl** [comment]
> > @aaronwi meanwhile, use SingleHTML save and use nightly... it really powerful for now at least..
> 
> a nightly alpha build is not suitable for daily use.

In theory, but the nightly on android is much more stable then the stable branch. I already use it for 2 years, never look back to stable.

**Comment by headbank** [comment]
> @aaronwi meanwhile, use SingleHTML save and use nightly... it really powerful for now at least..

What is SingleHTML? Would help if you explained this.

**Comment by benyaminl** [comment]
> > @aaronwi meanwhile, use SingleHTML save and use nightly... it really powerful for now at least..
> 
> What is SingleHTML? Would help if you explained this.

SingleHTML is a extension that save page as one single HTML. And you can open it also on android using HTML browser. 

![220619_085822](https://user-images.githubusercontent.com/2817332/174462751-7872b816-763c-4494-b569-e2f4d3f967cf.png)
and 

![220619_085920](https://user-images.githubusercontent.com/2817332/174464101-61925a32-4fbf-45b3-909f-132a60e165bb.png)

**Comment by headbank** [comment]
Right, so it's actually called [SingleFile](https://addons.mozilla.org/en-GB/firefox/addon/single-file/) ... correct name helps us find it ;) (There is a project called SingleHTML and I was trying to work out its relevance!)

This does sound like it could be a handy workaround in some cases. I expect it will have some issues with more Javascript-heavy pages though (depends on how sites use JS for rendering, and how the addon actually operates). Also I note from one of its reviews that it injects a "button" at the top of the rendered page (providing access to various metadata) which will often be undesirable when printing is the intended application. (Maybe this feature can be disabled in the addon's settings?)

It's worth mentioning in #219 too.

**Comment by gildas-lormeau** [comment]
> This does sound like it could be a handy workaround in some cases. I expect it will have some issues with more Javascript-heavy pages though (depends on how sites use JS for rendering, and how the addon actually operates). 
By default, SingleFile does not execute the JavaScript. It just saves a snapshot of the displayed page. The processing time depends more on the complexity of the CSS and HTML and the weight of embedded resources.
 
> Also I note from one of its reviews that it injects a "button" at the top of the rendered page (providing access to various metadata) which will often be undesirable when printing is the intended application. (Maybe this feature can be disabled in the addon's settings?)
I confirm it can be removed by disabling the option "User interface > display an infobar when viewing a saved page" in SingleFile.

For this use-case, I would also recommend to enable the option "Annotation editor > annotate the page before saving" in SingleFile. The annotation editor allows removing unwanted elements in the page and includes a "Print" button.

**Comment by Ansarmughal99** [comment]
Print page and view page as pdf is roll map for Firefox this year. Firefox team confirm this on twitter

**Comment by chriswitthuhn** [comment]
Any news on this? I see notes about road map for 2022 but we are a pretty good chunk of the way though the year and have seen nothing more specific than 2022.

**Comment by dahamsta** [comment]
Two and a half years now.

**Comment by benyaminl** [comment]
@Ansarmughal99 Would you please mention/put the link here? Thank you

**Comment by jr-johnson-support** [comment]
JR on the Customer Experience team here:

We have been tracking google play store reviews noting this request since May 2022 and seen 70 request thus far. 

A few more reviews noting this are usually added monthly.

**Comment by pyr0ball** [comment]
@jr-johnson-support would a few more help push the devs into considering this? I don't like doing that but if it'll get the features we want...

**Comment by headbank** [comment]
@pyr0ball I think the comment you reply to is itself the "push" you are looking for ... I doubt there is any formal pathway between what's happening in the Play reviews and the dev team, hence why those on that side are raising the issue here.

However please be aware that there is movement, however slow it seems, if you know where to look. Printing support has been dependent on Save To PDF support (because of how printing on Android works), which has been very busy lately and is finally getting ready for release - see [Bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1780639).

The [Bugzilla bug for printing](https://bugzilla.mozilla.org/show_bug.cgi?id=1659818) is quiet now but it will now be able to move forward, I'd suggest subscribing to it (and there is another ticket [here](https://github.com/mozilla-mobile/fenix/issues/13214) too) if you want up-to-date overview of progress.

I dunno where work is at on the printing UI (we don't see all that they do in public) but at least once Save To PDF is live, there will be a workaround by printing from a PDF app that does support printing.

**Comment by jr-johnson-support** [comment]
> @jr-johnson-support would a few more help push the devs into considering this? I don't like doing that but if it'll get the features we want...

Headbank explained things pretty well on this front. What I will add is this:

In general, a few more request for a specific feature is not what will get that feature over the line to get implemented. There is a multitude of reasons why work on a feature will happen or wont happen. As headbank noted,  there has not always been a formal pathway between various channels mobile users are posting and the dev team. My team is working towards improving visibility of the most common user issues and feedback across these channels. 

As noted above, there are lots of moving pieces and dependencies that mobile team has to work with. This can make it hard to work at the pace which some users want and expect.

**Comment by pyr0ball** [comment]
@headbank @jr-johnson-support 
Thank you both for your explanations! It's always nice to get a little insight and have a few more places to look for answers

**Comment by cpeterso** [comment]
It's not printing to paper, but "Save as PDF" will ship in Firefox 108: https://bugzilla.mozilla.org/show_bug.cgi?id=1780639

**Comment by headbank** [comment]
Also, there is now actually someone assigned to the [Bugzilla window.print() bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1659818) :)

Although at the same time it's been downgraded from "defect" to "enhancement" :(

**Comment by Matt-17** [comment]
Still no news?

**Comment by csadilek** [comment]
Moved to bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1808755

Change performed by the [Move to Bugzilla add-on](https://addons.mozilla.org/en-US/firefox/addon/move-to-bugzilla/).
